File: 114_Flatten_Binary_Tree_to_Linked_List.py
"""
Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
click to show hints.
"""

# Key Points: In pre-order place. Pre-order Traversal.
#
# Run time: 68 ms

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        def preorder_traversal(root):
            if not root:
                return []
            return [root] + preorder_traversal(root.left) + preorder_traversal(root.right)

        if not root:
            return root
        preorder = preorder_traversal(root)
        if len(preorder) == 1:
            return

        parent = root
        parent.left = None
        for node in preorder[1:]:
            parent.right = node
            node.left = None
            parent = node
        node.right = None

File: java/114_Flatten_Binary_Tree_to_Linked_List.java
// Given a binary tree, flatten it to a linked list in-place.

// For example,
// Given

//          1
//         / \
//        2   5
//       / \   \
//      3   4   6
// The flattened tree should look like:
//    1
//     \
//      2
//       \
//        3
//         \
//          4
//           \
//            5
//             \
//              6




// Runtime: 1 ms
// The helper function would return the last node of the flatten linked list

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        helper(root);
    }
    // return the last node of the flatten list
    private TreeNode helper(TreeNode root) {
        if (root == null)  return null;
        TreeNode leftNode = root.left;
        TreeNode rightNode = root.right;
        TreeNode leftList = helper(leftNode);
        TreeNode rightList = helper(rightNode);
        
        TreeNode tmp = root;
        if (leftList != null) {
            tmp.right = leftNode;
            tmp.left = null;
            tmp = leftList;
        }
        if (rightList != null) {
            tmp.right = rightNode;
            tmp.left = null;
            tmp = rightList;
        }
        tmp.left = null;
        tmp.right = null;
        return tmp;
    }
}